calyx-py/test/correctness/matrix.py

```diff
@@ -1,107 +1,128 @@
     A*B and writes this into C.
     """
 
-    logn = n.bit_length() 
-    
+    logn = n.bit_length()
+
     matmul = prog.component("matmul")
-    
-    # matrices
+
+    # The matrices are declared using a similar syntax.
     A = matmul.comb_mem_d2("A", 32, n, n, logn, logn, is_ref=True)
-    B = matmul.seq_mem_d2( "B", 32, n, n, logn, logn, is_ref=True)
-    C = matmul.seq_mem_d2( "C", 32, n, n, logn, logn, is_ref=True)
+    B = matmul.seq_mem_d2("B", 32, n, n, logn, logn, is_ref=True)
+    C = matmul.seq_mem_d2("C", 32, n, n, logn, logn, is_ref=True)
 
     mult = matmul.mult_pipe(32)
     add = matmul.add(32)
-
     acc = matmul.reg(32)
 
-    # iterators: i, j, k ∈ [0, n)
+    # Iterators: i, j, k ∈ [0, n)
     i = matmul.reg(3)
     j = matmul.reg(3)
     k = matmul.reg(3)
 
-    # matrix entries
-    a = matmul.reg(32) 
-    b = matmul.reg(32) 
-
-
-    zero_acc = matmul.reg_store(acc, 0) # acc := 0
-    zero_i = matmul.reg_store(i, 0)     # i := 0
-    zero_j = matmul.reg_store(j, 0)     # j := 0
-    zero_k = matmul.reg_store(k, 0)     # k := 0
-
-    cond_i = matmul.lt_use(i.out, n)    # i < n
-    cond_j = matmul.lt_use(j.out, n)    # j < n
-    cond_k = matmul.lt_use(k.out, n)    # k < n
-
-    read_A = matmul.mem_load_d2(A, i.out, k.out, a, "read_A") # a := A[i][k]
-    read_B = matmul.mem_load_d2(B, k.out, j.out, b, "read_B") # b := B[k][j]
-    
-    # C[i][j] := c
-    write_C = matmul.mem_store_d2(C, i.out, j.out, acc.out, "write") 
-   
-    # acc := acc + (a * b)
-    with matmul.group("upd") as upd:
-        # compute a*b
+    # A register to store an entry of matrix A. We won't need one for matrix B.
+    a = matmul.reg(32)
+
+    zero_acc = matmul.reg_store(acc, 0)  # acc := 0
+    zero_j = matmul.reg_store(j, 0)  # j := 0
+    zero_k = matmul.reg_store(k, 0)  # k := 0
+
+    cond_i = matmul.lt_use(i.out, n)  # i < n
+    cond_j = matmul.lt_use(j.out, n)  # j < n
+    cond_k = matmul.lt_use(k.out, n)  # k < n
+
+    # a := A[i][k]
+    load_A = matmul.mem_load_d2(A, i.out, k.out, a, "read_A")
+    # Latch B[k][j], so that we can later read `B.read_data` and get B[k][j].
+    # While `mem_load` works on combinational and sequential memories,
+    # `mem_latch` only works on combinational memories.
+    latch_B = matmul.mem_latch_d2(B, k.out, j.out, "read_B")
+
+    # C[i][j] := acc
+    write_C = matmul.mem_store_d2(C, i.out, j.out, acc.out, "write")
+    # We do not demonstrate it here, but storing into a combinational memory
+    # works identically. The `mem_store` method is overloaded.
+
+    # acc := acc + (a * b), where b is the value latched at B
+    with matmul.group("upd_acc") as upd_acc:
+        # Compute a*b
         mult.go = 1
         mult.left = a.out
-        mult.right = b.out
-        
-        # compute acc + (a*b)
-        add.left = mult.done @ mult.out
-        add.right = mult.done @ acc.out
-        
-        # store acc + (a*b) in acc
+        mult.right = B.read_data
```

Comment:
Exhibit B of the main point! `B.read_data` directly, as opposed to `b.out` where `b` is a silly register whose whole job is to hold the value for a beat.